dbms/src/Storages/KVStore/MultiRaft/RegionSerde.cpp

```diff
@@ -0,0 +1,279 @@
+    {
+        // Compat
+        return std::make_pair(RegionPersistFormat::EagerTruncate, sizeof(UInt64));
+    }
+
+    if (!maybe_flag.has_value())
+    {
+        auto size = readBinary2<UInt32>(buf);
+        return std::make_pair(RegionPersistFormat::Unrecognizable, size);
+    }
+    auto flag = maybe_flag.value();
+    UInt32 size = 0;
+    if (flag == RegionPersistFormat::Finished)
+    {
+        // size = 0
+    }
+    else
+    {
+        size = readBinary2<UInt32>(buf);
+    }
+    return std::make_pair(flag, size);
+}
+
+size_t writePersistExtension(WriteBuffer & wb, UInt32 flag_encoded, const char * data, UInt32 size)
+{
+    auto total_size = writeBinary2(flag_encoded, wb);
+    total_size += writeBinary2(size, wb);
+    wb.write(data, size);
+    total_size += size;
+    return total_size;
+}
+
+size_t writePersistExtension(WriteBuffer & wb, RegionPersistFormat flag, const char * data, UInt32 size)
+{
+    RUNTIME_CHECK(flag != RegionPersistFormat::Finished);
+    auto flag_encoded = magic_enum::enum_underlying(flag);
+    if (flag == RegionPersistFormat::EagerTruncate)
+    {
+        // Compat
+        auto total_size = writeBinary2(flag_encoded, wb);
+        wb.write(data, size);
+        total_size += size;
+        return total_size;
+    }
+    else
+    {
+        return writePersistExtension(wb, flag_encoded, data, size);
+    }
+}
+
+size_t writePersistExtensionSuffix(WriteBuffer & wb)
+{
+    return writeBinary2(magic_enum::enum_underlying(RegionPersistFormat::Finished), wb);
+}
+
+std::tuple<size_t, UInt64> Region::serialize(WriteBuffer & buf) const
+{
+    auto binary_version = Region::CURRENT_VERSION;
+    fiu_do_on(FailPoints::force_region_persist_version, {
+        if (auto v = FailPointHelper::getFailPointVal(FailPoints::force_region_persist_version); v)
+        {
+            binary_version = std::any_cast<UInt64>(v.value());
+            LOG_WARNING(
+                Logger::get(),
+                "Failpoint force_region_persist_version set region binary version, value={}",
+                binary_version);
+        }
+    });
+    size_t total_size = writeBinary2(binary_version, buf);
+    UInt64 applied_index = -1;
+
+    {
+        std::shared_lock<std::shared_mutex> lock(mutex);
+
+        // Serialize meta
+        const auto [meta_size, index] = meta.serialize(buf);
+        total_size += meta_size;
+        applied_index = index;
+
+        // Try serialize extra flags
+        if (binary_version >= 2)
+        {
+            static_assert(sizeof(eager_truncated_index) == sizeof(UInt64));
+            WriteBufferFromOwnString sub_buf;
+            auto sub_size = writeBinary2(eager_truncated_index, sub_buf);
+            RUNTIME_CHECK(sub_size, sizeof(UInt64));
+            auto s = sub_buf.releaseStr();
+            total_size += writePersistExtension(buf, RegionPersistFormat::EagerTruncate, s.data(), s.size());
+        }
```

Comment:
In the current branch, the binary format of extra flags is
```
| UINT32 |               UINT64 |
|  flags | eager_truncate_index |
```

https://github.com/pingcap/tiflash/blob/43f809fffde22d0af4c519be4546a5bf4dde30a2/dbms/src/Storages/KVStore/Region.cpp#L178-L185

But in this commit, the binary format is
```
|               UINT64 | UINT32 |
| eager_truncate_index | flag |
```

Reply:
It actually writes `sub_buf` rather than `buf`